crates/ruff_python_formatter/src/statement/stmt_if.rs

```diff
@@ -96,35 +48,56 @@ impl FormatNodeRule<StmtIf> for FormatStmtIf {
     }
 }
 
-fn else_if(or_else: &Suite) -> Option<IfOrElIf> {
-    if let [Stmt::If(if_stmt)] = or_else.as_slice() {
-        Some(IfOrElIf::ElIf(if_stmt))
-    } else {
-        None
-    }
-}
-
-enum IfOrElIf<'a> {
-    If(&'a StmtIf),
-    ElIf(&'a StmtIf),
-}
-
-impl<'a> IfOrElIf<'a> {
-    const fn statement(&self) -> &'a StmtIf {
-        match self {
-            IfOrElIf::If(statement) => statement,
-            IfOrElIf::ElIf(statement) => statement,
-        }
-    }
+/// Note that this implementation misses the leading newlines before the leading comments because
+/// it does not have access to the last node of the previous branch
+#[derive(Default)]
+pub struct FormatElifElseClause;
 
-    const fn keyword(&self) -> &'static str {
-        match self {
-            IfOrElIf::If(_) => "if",
-            IfOrElIf::ElIf(_) => "elif",
-        }
+impl FormatNodeRule<ElifElseClause> for FormatElifElseClause {
+    fn fmt_fields(&self, item: &ElifElseClause, f: &mut PyFormatter) -> FormatResult<()> {
+        format_elif_else_clause(item, f, None)
     }
+}
 
-    const fn is_elif(&self) -> bool {
-        matches!(self, IfOrElIf::ElIf(_))
+/// Extracted so we can implement `FormatElifElseClause` but also pass in `last_node` from
+/// `FormatStmtIf`
+fn format_elif_else_clause(
+    item: &ElifElseClause,
+    f: &mut PyFormatter,
+    last_node: Option<AnyNodeRef>,
+) -> FormatResult<()> {
+    let ElifElseClause {
+        range: _,
+        test,
+        body,
+    } = item;
+
+    let comments = f.context().comments().clone();
+    let trailing_colon_comment = comments.dangling_comments(item);
+    let leading_comments = comments.leading_comments(item);
+
+    if let Some(test) = test {
+        write!(
+            f,
+            [
+                leading_alternate_branch_comments(leading_comments, last_node),
+                text("elif"),
+                space(),
+                test.format().with_options(Parenthesize::IfBreaks),
+                text(":"),
+                trailing_comments(trailing_colon_comment),
+                block_indent(&body.format())
+            ]
+        )
+    } else {
+        write!(
+            f,
+            [
+                leading_alternate_branch_comments(leading_comments, last_node),
+                text("else:"),
+                trailing_comments(trailing_colon_comment),
+                block_indent(&body.format())
+            ]
+        )
     }
```

Comment:
Nit: I find it hard to identify what the difference is between the two branches if they alsoc contain the shared logic
```suggestion
    leading_alternate_branch_comments(leading_comments, last_node).fmt(f)?;
    
    if let Some(test) = test {
        write!(
            f,
            [
                text("elif"),
                space(),
                test.format().with_options(Parenthesize::IfBreaks),
                text(":")
            ]
        )
    } else {
    	text("else").fmt(f)?;
    }
    
    write!(f, [
                trailing_comments(trailing_colon_comment),
                block_indent(&body.format())
  	])
```

Reply:
i added this, but i find it confusing that one block is now doing up to the colon, then the next one has both the rest of the same line and the body lines